File: rules/culo.py
import pygame
import sys

class AI:
    def __init__(self):
        pass

class Game:
    def __init__(self,gamezone,actions):
        self.gamezone=gamezone
        self.actions=actions
        self.throws=[]
        self.ai=AI()
        self.caption="Culo"
        self.playzone=True
        self.deckdraws=[{"name":"Mazo para robar","numbers":["As",2,3,4,5,6,7,"J","Q","K"],"suits":["espadas","oros","bastos","copas"]}]
        self.down_func={
            pygame.K_DOWN   :   [pygame.display.toggle_fullscreen,[]] , \
            pygame.K_ESCAPE :   actions.exit, \
            pygame.K_F5     :   actions.show, \
            pygame.K_1      :   [actions.select,0], \
            pygame.K_2      :   [actions.select,1], \
            pygame.K_3      :   [actions.select,2], \
            pygame.K_4      :   [actions.select,3], \
            pygame.K_5      :   [actions.select,4], \
            pygame.K_6      :   [actions.select,5], \
            pygame.K_7      :   [actions.select,6], \
            pygame.K_8      :   [actions.select,7], \
            pygame.K_9      :   [actions.select,8], \
            pygame.K_0      :   [actions.select,9], \
            pygame.K_x      :   [actions.select,10], \
            pygame.K_c      :   [actions.select,11], \
            pygame.K_v      :   [actions.select,12], \
            pygame.K_b      :   [actions.select,13], \
            pygame.K_n      :   [actions.select,14], \
            pygame.K_m      :   [actions.select,15], \
            pygame.K_l      :   [actions.select,16], \
            pygame.K_k      :   [actions.select,17], \
            pygame.K_j      :   [actions.select,18], \
            pygame.K_h      :   [actions.select,19], \
            pygame.K_g      :   [actions.select,20], \
            pygame.K_f      :   [actions.select,21], \
            pygame.K_z      :   actions.clear_selection, \
            pygame.K_t      :   actions.throw_cards, \
            pygame.K_RETURN :   actions.throw_cards, \
            pygame.K_e      :   actions.end_turn, \
            pygame.K_s      :   actions.sort_by_points, \
            pygame.K_d      :   actions.draw_a_card, \
            pygame.K_p      :   actions.pass_turn
        }

    def points(self,number,suit):
        if number=="As":
            r=14
        elif number=="J":
            r=11
        elif number=="Q":
            r=12
        elif number=="K":
            r=13
        else:
            r=number
        return r
        
    def init(self):
        self.gamezone.deckdraws[0].shuffle()
        self.actions.deal(60) #repartir cartas    
        
    def pass_turn(self):
        if self.gamezone.pass_turns_counter==len(self.gamezone.players)-1:
            #dar el turno al jugador que tiro la ultima carta, no hace falta porque casualmente es el siguiente
            #self.gamezone.players.index( self.throws[len(self.throws)-1][0].previous_owner)
            
            self.actions.clear_playzone()
            del self.throws[:]

    def throw_cards(self,selection):
        self.actions.ending_turn()
        self.throws.append(selection)
            
    def throwable_selection(self,selection):
        #tiene que haber algo seleccionado
        if len(selection)==0:return False
        
        #obtenemos el numero y todos han de ser iguales
        number=selection[0].number
        for card in selection:
            if not card.number==number: return False
        
        #si se cumple lo anterior y no ha habido tiradas, cualquiera es valida
        if len(self.throws)==0:
            return True
        
        #si ya hay tiradas entonces tendran que comprobarse mas parametros 
        else:
            last_throw=self.throws[len(self.throws)-1]
            if len(selection)==len(last_throw):
                if selection[0].points>last_throw[0].points:
                    return True
            return False
            

    def new_turn(self):
        print "----------->----------->-------------->.>"
        if self.gamezone.player_with_turn>0:
            self.actions.sort_by_points()
                #last_throw=throws[len(throws)-1][0].number
       
            

File: rules/culo.lua
rules.caption = "Culo"
rules.descruption = "Game of Culo"
rules.playzone = 1

rules.deckdraws = {
    {
        name = "Mazo para robar",
        numbers = {"As", 2, 3, 4, 5, 6, 7, "J", "Q", "K"},
        suits = {"spades", "diamonds", "clubs", "hearts"}
    }
}

rules.keys_descriptions="\
F5:Reload  Esc:Exit  Down:Toggle Fullscreen  D:Draw a card \
[1-10]:Add to/Rem from Selection  Z:Clear Selection  T,Return:Throws \
E:End Turn  S:Sort P:Pass Turn F1-4:Choose User  F12:None User"

rules.down_func={
    K_DOWN   =   {"pygame.display.toggle_fullscreen()","local"} ,
    K_ESCAPE =   {"self.exit()","local"}, 
    K_F5     =   {"self.show()","local"}, 
    K_UP     =   {"self.showalt()","local"}, 
    K_z      =   {"self.clear_selection","global"}, 
    K_t      =   {"self.throw_cards","global"}, 
    K_RETURN =   {"self.throw_cards","global"}, 
    K_e      =   {"self.end_turn()","global"}, 
    K_s      =   {"self.sort('points')","local"}, 
    K_d      =   {"self.draw_a_card","global"}, 
    K_p      =   {"self.pass_turn","global"}, 

    K_1       =   game.prueba1, 
    K_2       =   game.prueba2, 

    K_F1      =   game.F1, 
    K_F2      =   game.F2, 
    K_F3      =   game.F3, 
    K_F4      =   game.F4, 
    K_F12     =   game.F12, 
}
 
function rules:points(number,suit)
    if number=="As" then
        r=14
    elseif number=="J" then
        r=11
    elseif number=="Q" then
        r=12
    elseif number=="K" then
        r=13
    else
        r=number
    end
    return r
end
        
function rules:new_round()
    getattr(game.deckdraws[0], 'shuffle')()
    game.deal(60) --repartir cartas    
end
    
function rules:pass_turn()
    if game.pass_turns_counter==len(game.players)-1 then
        --dar el turno al jugador que tiro la ultima carta, no hace falta porque casualmente es el siguiente
        --#self.gamezone.players.index( self.throws[len(self.throws)-1][0].previous_owner)

        game.clear_playzone()

        --del self.throws[:]
        --for k in self.throws do table.remove(self.throws,k) end

        self.throws = {}

    end
end

function rules:throw_cards(selection)
    game.ending_turn()
    table.insert(self.throws, selection)
end

function rules:throwable_selection(selection)
    --tiene que haber algo seleccionado
    if len(selection)==0 then return false end
    
    --obtenemos el numero y todos han de ser iguales
    
    number=selection[1].number

    --for card in selection
    --    if not card.number==number then return false end
    
    for k,card in pairs(selection) do 
        if not card.number == number then return false end
    end
    
    --si se cumple lo anterior y no ha habido tiradas, cualquiera es valida
    if len(self.throws) == 0 then
        return true
    --si ya hay tiradas entonces tendran que comprobarse mas parametros 
    else
        last_throw=self.throws[len(self.throws)]
        if len(selection)==len(last_throw) then
            if selection[1].points > last_throw[1].points then
                return true
            end
        end
        return false
    end
end


--function rules:is_round_finished() return true end

